apiserver/facades/client/application/application.go

```diff
@@ -2199,3 +2275,77 @@ func (api *APIBase) ApplicationsInfo(in params.Entities) (params.ApplicationInfo
 	}
 	return params.ApplicationInfoResults{out}, nil
 }
+
+// lxdCharmProfiler massages a *state.Charm into a LXDProfiler
+// inside of the core package.
+type lxdCharmProfiler struct {
+	Charm Charm
+}
+
+// LXDProfile implements core.lxdprofile.LXDProfiler
+func (p lxdCharmProfiler) LXDProfile() lxdprofile.LXDProfile {
+	if p.Charm == nil {
+		return nil
+	}
+	if profiler, ok := p.Charm.(charm.LXDProfiler); ok {
+		return profiler.LXDProfile()
+	}
+	return nil
+}
+
+// AgentTools is a point of use agent tools requester.
+type AgentTools interface {
+	AgentTools() (*tools.Tools, error)
+}
+
+// ModelConfig is a point of use model configuration.
+type ModelConfig interface {
+	AgentVersion() (version.Number, error)
+}
+
+var (
+	// ErrInvalidAgentVersions is a sentinal error for when we can no longer
+	// upgrade juju using 2.5.x agents with 2.6 or greater controllers.
+	ErrInvalidAgentVersions = errors.Errorf(
+		"Unable to upgrade LXDProfile charms with the current model version. " +
+			"Please run juju upgrade-juju to upgrade the current model to match your controller.")
+)
+
+func getAgentToolsVersion(agentTools AgentTools) (version.Number, error) {
+	tools, err := agentTools.AgentTools()
+	if err != nil {
+		return version.Zero, err
+	}
+	return tools.Version.Number, nil
+}
+
+func getModelConfigVersion(modelConfig ModelConfig) (version.Number, error) {
+	agent, err := modelConfig.AgentVersion()
+	if err != nil {
+		return version.Zero, err
+	}
+	return agent, nil
+}
+
+func validateAgentVersions(application Application, modelConfig ModelConfig) error {
+	// The epoch is set like this, because beta tags are less than release tags.
+	// So 2.6-beta1.1 < 2.6.0, even though the patch is greater than 0. To
+	// prevent the miss-match, we add the upper epoch limit.
+	epoch := version.Number{Major: 2, Minor: 5, Patch: math.MaxInt32}
+
+	ver, err := getAgentToolsVersion(application)
+	if ver.Compare(epoch) >= 0 || errors.IsNotFound(err) {
```

Comment:
don't you need to check the error *before* you treat the version as valid?
we probably also want to trace the error if we get a failure that we don't understand, which seems like something:
if err != nil && !errors.IsNotFound(err) {
    return errors.Trace(err)
}

I do believe application.AgentTools is only valid for CAAS models, because there you have 1 agent for the application, but no agents for the Units. However on IAAAS models, there is only the Unit.AgentTools or the Model agent tools. I guess maybe that is your IsNotFound check?

maybe just a comment explaining what we are looking for. Do we even need to bother with the Application version rather than just checking the ModelConfig desired agent version? (arguably you might want to check on the per-unit level, as that is the actual version of the agent that is running, but if you look at the ModelConfig, it is telling you the intent is to move to that version eventually, which seems sufficient for a pre-check like this.)

Reply:
> maybe just a comment explaining what we are looking for. Do we even need to bother with the Application version rather than just checking the ModelConfig desired agent version? (arguably you might want to check on the per-unit level, as that is the actual version of the agent that is running, but if you look at the ModelConfig, it is telling you the intent is to move to that version eventually, which seems sufficient for a pre-check like this.)

This is exactly what I'm think and was hoping you'd agree....